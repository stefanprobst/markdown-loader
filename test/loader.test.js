/** @jest-environment node */

const compiler = require('./utils/compiler.js')

const fixture = 'fixtures/test.md'

it('Transforms Markdown to JavaScript', async () => {
  const stats = await compiler(fixture)
  const { source: output } =
    stats.toJson().modules.find((module) => module.name.endsWith(fixture)) || {}

  expect(output).toMatchSnapshot()
})

it('Optionally allows inline html', async () => {
  const stats = await compiler(fixture, { allowDangerousHtml: true })
  const { source: output } =
    stats.toJson().modules.find((module) => module.name.endsWith(fixture)) || {}

  expect(output).toMatchSnapshot()
})

it('Accepts additional remark plugins', async () => {
  const remarkPlugins = [require('remark-footnotes')]
  const stats = await compiler(fixture, { remarkPlugins })
  const { source: output } =
    stats.toJson().modules.find((module) => module.name.endsWith(fixture)) || {}

  expect(output).toMatchSnapshot()
})

it('Accepts additional remark plugins with plugin options', async () => {
  const remarkPlugins = [[require('remark-footnotes'), { inlineNotes: true }]]
  const stats = await compiler(fixture, { remarkPlugins })
  const { source: output } =
    stats.toJson().modules.find((module) => module.name.endsWith(fixture)) || {}

  expect(output).toMatchSnapshot()
})

it('Allows importing html and frontmatter', async () => {
  const stats = await compiler(fixture)
  const { source: output } =
    stats.toJson().modules.find((module) => module.name.endsWith(fixture)) || {}

  // esm to commonjs exports
  const { transform } = require('@babel/core')
  const { code } = await transform(output, {
    plugins: [['@babel/plugin-transform-modules-commonjs', { strict: true }]],
  })

  // load module from string
  const Module = require('module')
  const m = new Module()
  m._compile(code, '')
  const { data, default: html } = m.exports

  // render html
  const { createElement } = require('react')
  const renderer = require('react-test-renderer')
  const render = renderer.create(
    createElement('article', { dangerouslySetInnerHTML: html }),
  )

  expect(data).toMatchInlineSnapshot(`
    Object {
      "title": "Test document",
    }
  `)
  expect(render.toJSON()).toMatchInlineSnapshot(`
    <article
      dangerouslySetInnerHTML="<h2>Title</h2>
    <p>Some <strong>important</strong> text![^1] See <a href=\\"https://example.com\\">here</a> for more.^[inline
    note] Please ignore inline html.</p>
    <pre><code class=\\"language-js\\">const message = 'This is a fenced code block.'
    </code></pre>
    <p>[^1]: Well, not so important after all.</p>"
    />
  `)
})
