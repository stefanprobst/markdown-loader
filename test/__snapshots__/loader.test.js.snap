// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accepts additional remark plugins 1`] = `
"const data = {\\"title\\":\\"Test document\\"}
export { data }
export default \\"<h2>Title</h2>\\\\n<p>Some <strong>important</strong> text!<sup id=\\\\\\"fnref-1\\\\\\"><a href=\\\\\\"#fn-1\\\\\\" class=\\\\\\"footnote-ref\\\\\\">1</a></sup> See <a href=\\\\\\"https://example.com\\\\\\">here</a> for more.^[inline\\\\nnote] Please ignore inline html.</p>\\\\n<pre><code class=\\\\\\"language-js\\\\\\">const message = 'This is a fenced code block.'\\\\n</code></pre>\\\\n<div class=\\\\\\"footnotes\\\\\\">\\\\n<hr>\\\\n<ol>\\\\n<li id=\\\\\\"fn-1\\\\\\">Well, not so important after all.<a href=\\\\\\"#fnref-1\\\\\\" class=\\\\\\"footnote-backref\\\\\\">↩</a></li>\\\\n</ol>\\\\n</div>\\""
`;

exports[`Accepts additional remark plugins with plugin options 1`] = `
"const data = {\\"title\\":\\"Test document\\"}
export { data }
export default \\"<h2>Title</h2>\\\\n<p>Some <strong>important</strong> text!<sup id=\\\\\\"fnref-1\\\\\\"><a href=\\\\\\"#fn-1\\\\\\" class=\\\\\\"footnote-ref\\\\\\">1</a></sup> See <a href=\\\\\\"https://example.com\\\\\\">here</a> for more.<sup id=\\\\\\"fnref-2\\\\\\"><a href=\\\\\\"#fn-2\\\\\\" class=\\\\\\"footnote-ref\\\\\\">2</a></sup> Please ignore inline html.</p>\\\\n<pre><code class=\\\\\\"language-js\\\\\\">const message = 'This is a fenced code block.'\\\\n</code></pre>\\\\n<div class=\\\\\\"footnotes\\\\\\">\\\\n<hr>\\\\n<ol>\\\\n<li id=\\\\\\"fn-1\\\\\\">Well, not so important after all.<a href=\\\\\\"#fnref-1\\\\\\" class=\\\\\\"footnote-backref\\\\\\">↩</a></li>\\\\n<li id=\\\\\\"fn-2\\\\\\">inline\\\\nnote<a href=\\\\\\"#fnref-2\\\\\\" class=\\\\\\"footnote-backref\\\\\\">↩</a></li>\\\\n</ol>\\\\n</div>\\""
`;

exports[`Optionally allows inline html 1`] = `
"const data = {\\"title\\":\\"Test document\\"}
export { data }
export default \\"<h2>Title</h2>\\\\n<p>Some <strong>important</strong> text![^1] See <a href=\\\\\\"https://example.com\\\\\\">here</a> for more.^[inline\\\\nnote] Please ignore inline <code>html</code>.</p>\\\\n<pre><code>const message = 'This is a fenced code block.'\\\\n</code></pre>\\\\n<p>[^1]: Well, not so important after all.</p>\\""
`;

exports[`Transforms Markdown to JavaScript 1`] = `
"const data = {\\"title\\":\\"Test document\\"}
export { data }
export default \\"<h2>Title</h2>\\\\n<p>Some <strong>important</strong> text![^1] See <a href=\\\\\\"https://example.com\\\\\\">here</a> for more.^[inline\\\\nnote] Please ignore inline html.</p>\\\\n<pre><code class=\\\\\\"language-js\\\\\\">const message = 'This is a fenced code block.'\\\\n</code></pre>\\\\n<p>[^1]: Well, not so important after all.</p>\\""
`;

exports[`format: plaintext transforms markdown to plaintext 1`] = `
"const data = {\\"title\\":\\"Test document\\"}
export { data }
export default \\"\\\\n\\\\nTitle\\\\n\\\\nSome important text^1 See here for more.^inline\\\\nnote Please ignore inline html.\\\\n\\\\n\\\\n\\\\n^1: Well, not so important after all.\\\\n\\""
`;
